pkg/crd/crd.go

```diff
@@ -0,0 +1,136 @@
+package crd
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	apiextv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	clientv1 "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset/typed/apiextensions/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+var (
+	waitDuration = 1 * time.Minute
+	waitInterval = 500 * time.Millisecond
+)
+
+// BatchCreateCRDs will create or update the list of crds if they do not exist.
+func BatchCreateCRDs(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface, crds []*apiextv1.CustomResourceDefinition) error {
+	existingCRDs, err := getExistingCRD(ctx, crdClient)
+	if err != nil {
+		// we do not need to return this error because we can still attempt to create the CRDs if the list fails
+		logrus.Warnf("unable to check for existing CRDs: %s", err.Error())
+	}
+
+	// ensure each CRD exist
+	for _, crd := range crds {
+		err = ensureCRD(ctx, crd, crdClient, existingCRDs)
+		if err != nil {
+			return err
+		}
+	}
+
+	existingCRDs, err = getExistingCRD(ctx, crdClient)
+	if err != nil {
+		return fmt.Errorf("unable to check for existing CRDs: %w", err)
+	}
+	// ensure each CRD is ready
+	for _, crd := range crds {
+		if existingCRD, ok := existingCRDs[crd.Name]; ok && crdIsReady(existingCRD) {
+			continue
+		}
+		// wait for the CRD to be ready
+		err := waitCRD(ctx, crd.Name, crdClient)
+		if err != nil {
+			return fmt.Errorf("failed waiting on CRD '%s': %w", crd.Name, err)
+		}
+	}
+
+	return nil
+}
+
+// getExistingCRD returns a map of all CRD resource on the cluster.
+func getExistingCRD(ctx context.Context, crdClient clientv1.CustomResourceDefinitionInterface) (map[string]*apiextv1.CustomResourceDefinition, error) {
+	storedCRDs, err := crdClient.List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to list crds: %w", err)
+	}
+
+	// convert existingCRDs to a map for faster lookup
+	existingCRDs := map[string]*apiextv1.CustomResourceDefinition{}
+	for i := range storedCRDs.Items {
+		crd := &storedCRDs.Items[i]
+		existingCRDs[crd.Name] = crd
+	}
+	return existingCRDs, nil
+}
+
+// ensureCRD checks if there is an existing CRD that matches if not it will attempt to create or update the CRD.
+func ensureCRD(ctx context.Context, crd *apiextv1.CustomResourceDefinition, crdClient clientv1.CustomResourceDefinitionInterface, existingCRDs map[string]*apiextv1.CustomResourceDefinition) error {
+	existingCRD, exist := existingCRDs[crd.Name]
+	if !exist {
+		logrus.Infof("Creating embedded CRD %s", crd.Name)
+		_, err := crdClient.Create(ctx, crd, metav1.CreateOptions{})
+		if err == nil {
+			return nil
+		}
+		if !apierrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create crd '%s': %w", crd.Name, err)
+		}
+		// item was not in initial list but does exist so we will attempt to get the latest resourceVersion
```

Comment:
In what situation would this happen? Where `crdClient.List` would not contain something that could be gotten via `crdClient.Get`?

Reply:
2 scenarios, both of which don't get hit right now.

1. Allow for a label selector in the List so we do not have to pull every CRD we may not care about.
I will update the BatchCreate call to take one in.

2. If the user does not have list access the logic can still be performed if you have the get verb.
This does not work right now because of your previous comment where I am returning the error.